node/src/remote/client.ts

```diff
@@ -12,13 +12,98 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import axios, { type AxiosResponse } from 'axios'
+import axios, { type AxiosResponse, type ResponseType } from 'axios'
 
 import { tableFromIPC, type Table as ArrowTable } from 'apache-arrow'
 
+import { type RemoteResponse, type RemoteRequest, Method } from '../middleware'
+
+interface HttpLancedbClientMiddleware {
+  onRemoteRequest(
+    req: RemoteRequest,
+    next: (req: RemoteRequest) => Promise<RemoteResponse>,
+  ): Promise<RemoteResponse>
+}
+
+/**
+ * Invoke the middleware chain and at the end call the remote endpoint
+ */
+async function callWithMiddlewares (
+  req: RemoteRequest,
+  middlewares: HttpLancedbClientMiddleware[],
+  opts?: MiddlewareInvocationOptions
+): Promise<RemoteResponse> {
+  async function call (
+    i: number,
+    req: RemoteRequest
+  ): Promise<RemoteResponse> {
+    // if we have reached the end of the middleware chain, make the request
+    if (i > middlewares.length) {
+      let res
+      if (req.method === Method.POST) {
+        res = await axios.post(
+          req.uri,
+          req.data,
+          {
+            headers: { ...req.headers },
+            params: req.params,
+            responseType: opts?.responseType,
+            timeout: 10000
+          }
+        )
+      } else {
+        res = await axios.get(
+          req.uri,
+          {
+            headers: { ...req.headers },
+            params: req.params,
+            timeout: 10000
+          }
+        )
+      }
+
+      return toLanceRes(res)
+    }
+
+    // call next middleware in chain
+    return await middlewares[i - 1].onRemoteRequest(
+      req,
+      async (req) => {
+        return await call(i + 1, req)
```

Comment:
The indexing here seems off.  Should this be...

```
async func call(...) {
  if (i >= middlewares.length) {
    ...
  } else {
    middlewares[i].onRemoteRequest(..., async (req) => { return await call(i + 1, req); }
  }
return await call(0, req);
```

Reply:
This actually does work. For example if there are 2 middlewares [A, B]

When `call` is invoked for the first time (on line 77): `i = 1`, so we invoke the `i-1 = 0th = A` middleware's `onRemoteRequest` callback on line 69. Then when that middleware calls `next`, it invokes `call` for the second time on 72 with `i = i+1 = 2`

When `call` is invoked for the 2nd time, `i = 2` so we invoke the `i-1 = 1st = B` middleware's `onRemoteRequest` callback, and then when B middleware calls `next` it invokes `call` for the 3rd time with `i = i+1 = 3`

When `call` is invoked for the 3rd time, we get `i > middlewares.length` on line 41 and we make the request.